### DISCUSSION TITLE
"atomic energy" are not same as I defined in input?

### DISCUSSION THREAD

**Discussion by jtyang-chem** [deepmodeling/deepmd-kit#1443] [Q&A] [answered]
I did a train with `atom_ener`  specified, then I freeze and got a model, but the strange thing is, if I make an AB 2 atom  system and let the distance be like 20.0 (out of cut-range), the Ea, Eb I got from the model are not equal to the values I specified, and total energy is Ea+Eb but not what I expect.

Why?

**Comment by njzjz** [comment, marked as answer]
Did you set descriptors' `set_davg_zero` to `true`?

**Reply by jtyang-chem** [reply]
No, I didn't

**Reply by njzjz** [reply]
> No, I didn't

You need to set it to true to make atom_ener work.

**Reply by jtyang-chem** [reply]
Well, now I see I made a mistake, and what actually happend in my train?
If it still obey the eqution in [this post](https://github.com/deepmodeling/deepmd-kit/discussions/1357):
> ![](https://render.githubusercontent.com/render/math?math=E%5Ei%20%3D%20%5Cmathcal%7BN%7D(%5Cmathcal%7BD%7D%5Ei)%20-%20%5Cmathcal%7BN%7D(%5Cmathbf%7BD%7D%5Ei_0)%20%2B%20E%5Ei_0)

Then it should be `E^a= N(D^a_0)- N(D^a_0) + E^a_0`, I'll still got the vaule I specified before.

If it makes no differece, why did I see a different result when `atom_ener` was specified while `set_davg_zero` was as default?

**Reply by njzjz** [reply]
This equation looks good, but the actual behavior that the program does is `E^a= N(D)- N(D^0) + E^0`, minus `N(D^0)` instead, for better performance. Calculating `D^a_0` will cost extra time.

**Reply by jtyang-chem** [reply]
Yes, that's where the problem is, what is the difference between a `D` with only the atom itself in cut-range and `D_0` and vector **0**?

**Reply by njzjz** [reply]
The environment matrix `R` will be normalized. Thus, `D` will be also not zero.

https://github.com/deepmodeling/deepmd-kit/blob/bcfb127c4785e5f0783b3264acf31cba58bd1ba9/source/lib/src/prod_env_mat.cc#L80

**Reply by jtyang-chem** [reply]
As you said, I guess, when `set_davg_zero` is true, the D_0 in the equation is replaced with **0**.
But, when `set_davg_zero` is false, `D` and `D_0` will be calculated and normalized. 
My question is, why `D^a` with only one "a" atom itself in cut-range(without any neighbor) is not equal to the `D^a_0`?

**Reply by njzjz** [reply]
The normalization also takes effects on zero placeholders.

**Reply by njzjz** [reply]
When the `sel` is larger than the actual number of neighbors, the environment matrix will be filled with zero for these atoms that do not actually exist. This is helpful for occupying memory, as during training or MD simulations, we cannot expect the number of neighbors keeps the same.

**Reply by jtyang-chem** [reply]
OK, but still the question, when you use 0 inputs to calculate `D` (in other words, without any neighbors), it should be equal to `D_0`, right?
Or am I wrong? then what is the real meaning of `D_0`?

**Reply by njzjz** [reply]
`D_0` is a [zero matrix](https://en.wikipedia.org/wiki/Zero_matrix).

**Reply by jtyang-chem** [reply]
Excuse me for continuing the question,
The difference between `D` without neighbors and `D_0`, with `set_davg_zero` as false:
`D_0` is always zeros matrix, right?
And `D`, after checked the code, `avg` can be 0 or 1e-15, `std` will be 1e-2, so the elements of `D` will  be -1e-13 or 0, not a zeros matrix, but very close, right?
`D` close to `D_0`, however, the result I got from A-B system, showed `N(D)-N(D_0)` not close to 0, it can be very big, even about -100, could you explain this?

**Reply by njzjz** [reply]
No, `avg` and `std` are used to normalize `R` but not `D`.

See [the equations](https://docs.deepmodeling.org/projects/deepmd/en/v2.0.3/api_py/deepmd.descriptor.html#deepmd.descriptor.se_a.DescrptSeA).

**Reply by jtyang-chem** [reply]
Even though, how to explain the big difference?

**Reply by njzjz** [reply]
`avg` and `std` are computed based on training data. They are not fixed values.

**Reply by jtyang-chem** [reply]
Sure, they are not,  but for calculation of no-neighbor system, the `avg` and `std` are still from the training with-neighbor training  data?

**Reply by njzjz** [reply]
They are average of all your data. Do you only provide no-neighbor system?

**Reply by jtyang-chem** [reply]
No. That's explained the difference, thx for your patience.
